lib/spack/spack/solver/concretize.lp

```diff
@@ -7,31 +7,114 @@
+:- attr("node_platform", PackageNode, _), not attr("node", PackageNode).
+:- attr("node_os", PackageNode, _), not attr("node", PackageNode).
+:- attr("node_target", PackageNode, _), not attr("node", PackageNode).
+:- attr("node_compiler_version", PackageNode, _, _), not attr("node", PackageNode).
+:- attr("variant_value", PackageNode, _, _), not attr("node", PackageNode).
+:- attr("node_flag_compiler_default", PackageNode), not attr("node", PackageNode).
+:- attr("node_flag", PackageNode, _, _), not attr("node", PackageNode).
+:- attr("node_flag_source", PackageNode, _, _), not attr("node", PackageNode).
+:- attr("no_flags", PackageNode, _), not attr("node", PackageNode).
+:- attr("external_spec_selected", PackageNode, _), not attr("node", PackageNode).
+:- attr("depends_on", ParentNode, _, _), not attr("node", ParentNode).
+:- attr("depends_on", _, ChildNode, _), not attr("node", ChildNode).
+:- attr("node_flag_source", ParentNode, _, _), not attr("node", ParentNode).
+:- attr("node_flag_source", _, _, ChildNode), not attr("node", ChildNode).
+
+:- attr("virtual_node", VirtualNode), not provider(_, VirtualNode).
+:- provider(_, VirtualNode), not attr("virtual_node", VirtualNode).
+:- provider(PackageNode, _), not attr("node", PackageNode).
+
+:- attr("root", node(ID, PackageNode)), ID > main_node_id.
+
+% Nodes in the "root" unification set cannot depend on non-root nodes if the dependency is "link" or "run"
+:- attr("depends_on", node(main_node_id, Package), node(ID, _), "link"), ID != main_node_id, unification_set("root", node(main_node_id, Package)).
+:- attr("depends_on", node(main_node_id, Package), node(ID, _), "run"),  ID != main_node_id, unification_set("root", node(main_node_id, Package)).
+
+% Rules on "unification sets", i.e. on sets of nodes allowing a single configuration of any given package
+unify(SetID, PackageName) :- unification_set(SetID, node(_, PackageName)).
+:- 2 { unification_set(SetID, node(_, PackageName)) }, unify(SetID, PackageName).
+
+unification_set("root", PackageNode) :- attr("root", PackageNode).
+unification_set(SetID, ChildNode) :- attr("depends_on", ParentNode, ChildNode, Type), Type != "build", unification_set(SetID, ParentNode).
+
+unification_set(("build", node(X, Child)), node(X, Child))
+  :- attr("depends_on", ParentNode, node(X, Child), Type),
+     Type == "build",
+     SetID != "generic_build",
+     multiple_unification_sets(Child),
+     unification_set(SetID, ParentNode).
+
+unification_set("generic_build", node(X, Child))
+  :- attr("depends_on", ParentNode, node(X, Child), Type),
+     Type == "build",
+     not multiple_unification_sets(Child),
+     unification_set(_, ParentNode).
+
+% Any dependency of type "build" in a unification set that is not "root", stays in that unification set
+unification_set(SetID, ChildNode) :- attr("depends_on", ParentNode, ChildNode, Type), Type == "build", SetID == "generic_build", unification_set(SetID, ParentNode).
```

Comment:
can these long lines be split up? I think this file should have a column limit of 100 just like the others... we just don't have a formatter for it.

Reply:
Done in 5408d69abd1fae5dabb8efaf2e7fae219e75f1f2